src/StoreMeta.cc

```diff
@@ -9,172 +9,147 @@
-        return false;
-    }
+#include <limits>
 
-    /* Unused in any current squid code */
-    if (type == STORE_META_KEY_URL ||
-            type == STORE_META_KEY_SHA ||
-            type == STORE_META_HITMETERING ||
-            type == STORE_META_VALID) {
-        debugs(20, DBG_CRITICAL, "ERROR: Obsolete and unused type (" << type << ") in disk metadata");
-        return false;
-    }
+namespace Store {
 
-    return true;
+/// Whether the given raw swap meta field type represents a type that we should
+/// inform the admin about (if found in a store) but can otherwise ignore.
+inline constexpr
+bool
+DeprecatedSwapMetaType(const RawSwapMetaType type)
+{
+    return
+        type == 1 || // STORE_META_KEY_URL
+        type == 2 || // STORE_META_KEY_SHA
+        type == 6 || // STORE_META_HITMETERING (RFC 2227)
+        type == 7 || // STORE_META_VALID
+        false;
 }
 
-const int StoreMeta::MinimumTLVLength = 0;
-const int StoreMeta::MaximumTLVLength = 1 << 16;
-
+/// Whether the given raw swap meta field type represents a type that we should
+/// ignore without informing the admin.
+inline constexpr
 bool
-StoreMeta::validLength(int aLength) const
+ReservedSwapMetaType(const RawSwapMetaType type)
 {
-    static const Range<int> TlvValidLengths = Range<int>(StoreMeta::MinimumTLVLength, StoreMeta::MaximumTLVLength);
-    if (!TlvValidLengths.contains(aLength)) {
-        debugs(20, DBG_CRITICAL, MYNAME << ": insane length (" << aLength << ")!");
-        return false;
-    }
-
-    return true;
+    return
+        type == 11 || // STORE_META_STOREURL (Store-ID URL)
+        type == 12 || // STORE_META_VARY_ID (linking HTTP Vary variants)
+        false;
 }
 
-StoreMeta *
-StoreMeta::Factory (char type, size_t len, void const *value)
+/// Whether the given raw swap meta field type can be safely ignored.
+/// \sa HonoredSwapMetaType()
+inline constexpr
+bool
+IgnoredSwapMetaType(const RawSwapMetaType type)
 {
-    if (!validType(type))
-        return NULL;
-
-    StoreMeta *result;
-
-    switch (type) {
-
-    case STORE_META_KEY:
-        result = new StoreMetaMD5;
-        break;
+    return DeprecatedSwapMetaType(type) || ReservedSwapMetaType(type);
+}
 
-    case STORE_META_URL:
-        result = new StoreMetaURL;
-        break;
+/// Whether we store the given swap meta field type (and also interpret the
+/// corresponding swap meta field when the Store loads it). Matches all
+/// SwapMetaType enum values except for the never-stored/loaded STORE_META_VOID.
+inline constexpr
+bool
+HonoredSwapMetaType(const RawSwapMetaType type)
+{
+    return 0 < type && type <= SwapMetaTypeMax && !IgnoredSwapMetaType(type);
+}
 
-    case STORE_META_STD:
-        result = new StoreMetaSTD;
-        break;
+static_assert(SwapMetaTypeMax <= std::numeric_limits<RawSwapMetaType>::max(),
+              "RawSwapMetaType fits all SwapMetaType values");
```

Comment:
The placement of this is a bit odd. It should go at the top of this file. Or better: in the header where SwapMetaTypeMax value is defined.